File: snippets/java/LauncherModificationExamples.java
package dev.hmclce.sdk.examples;

import javafx.application.Platform;
import javafx.scene.control.Alert;
import org.jackhuang.hmcl.game.HMCLGameRepository;
import org.jackhuang.hmcl.plugin.PluginContext;
import org.jackhuang.hmcl.setting.GameDirectoryManager;
import org.jackhuang.hmcl.ui.Controllers;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.stream.Collectors;

public final class LauncherModificationExamples {
    private LauncherModificationExamples() {
    }

    public static void showDialog(String message) {
        Controllers.dialog(message, "Plugin SDK");
    }

    public static void showJavaFxAlert(PluginContext context) {
        Platform.runLater(() -> {
            Alert alert = new Alert(Alert.AlertType.INFORMATION);
            alert.initOwner(context.getPrimaryStage());
            alert.setTitle("Plugin SDK");
            alert.setHeaderText("JavaFX Alert");
            alert.setContentText("Plugins can create normal JavaFX UI.");
            alert.show();
        });
    }

    public static void changeWindowTitle(PluginContext context, String title) {
        Platform.runLater(() -> context.getPrimaryStage().setTitle(title));
    }

    public static String listGameInstances() {
        HMCLGameRepository repository = GameDirectoryManager.getSelectedRepository();
        return repository.getVersions().stream()
                .map(version -> version.getId())
                .collect(Collectors.joining("\n"));
    }

    public static void navigateToSettings() {
        Controllers.navigate(Controllers.getSettingsPage());
    }

    public static void navigateToDownloadPage() {
        Controllers.getDownloadPage().showGameDownloads();
        Controllers.navigate(Controllers.getDownloadPage());
    }

    public static void writePluginData(PluginContext context, String fileName, String content) throws Exception {
        Path target = context.getPluginDirectory().resolve(fileName);
        Files.createDirectories(target.getParent());
        Files.writeString(target, content);
    }

    public static String readPluginData(PluginContext context, String fileName) throws Exception {
        Path target = context.getPluginDirectory().resolve(fileName);
        return Files.exists(target) ? Files.readString(target) : "";
    }
}

File: snippets/kotlin/LauncherModificationExamples.kt
package dev.hmclce.sdk.examples

import javafx.application.Platform
import javafx.scene.control.Alert
import org.jackhuang.hmcl.plugin.PluginContext
import org.jackhuang.hmcl.setting.GameDirectoryManager
import org.jackhuang.hmcl.ui.Controllers
import java.nio.file.Files

object LauncherModificationExamples {
    fun showDialog(message: String) {
        Controllers.dialog(message, "Plugin SDK")
    }

    fun showJavaFxAlert(context: PluginContext) {
        Platform.runLater {
            Alert(Alert.AlertType.INFORMATION).apply {
                initOwner(context.primaryStage)
                title = "Plugin SDK"
                headerText = "JavaFX Alert"
                contentText = "Plugins can create normal JavaFX UI."
                show()
            }
        }
    }

    fun changeWindowTitle(context: PluginContext, title: String) {
        Platform.runLater { context.primaryStage.title = title }
    }

    fun listGameInstances(): String {
        val repository = GameDirectoryManager.getSelectedRepository()
        return repository.versions.joinToString("\n") { it.id }
    }

    fun navigateToSettings() {
        Controllers.navigate(Controllers.getSettingsPage())
    }

    fun navigateToDownloadPage() {
        Controllers.getDownloadPage().showGameDownloads()
        Controllers.navigate(Controllers.getDownloadPage())
    }

    fun writePluginData(context: PluginContext, fileName: String, content: String) {
        val target = context.pluginDirectory.resolve(fileName)
        Files.createDirectories(target.parent)
        Files.writeString(target, content)
    }

    fun readPluginData(context: PluginContext, fileName: String): String {
        val target = context.pluginDirectory.resolve(fileName)
        return if (Files.exists(target)) Files.readString(target) else ""
    }
}
